File: 1 semester/Task 7-2/Task 7-2/stack.cpp
#include <iostream>
#include "stack.h"

struct StackElement
{
	int value;
	StackElement *next;
};

struct Stack
{
	StackElement *top = nullptr;
};

Stack* createStack()
{
	Stack *stack = new Stack;
	return stack;
}

void push(Stack *stack, int value)
{
	StackElement *stackElement = new StackElement;
	stackElement->value = value;
	stackElement->next = stack->top;
	stack->top = stackElement;
}

int pop(Stack *stack)
{
	if (stack->top == nullptr)
	{
		return 1;
	}

	int result = stack->top->value;
	StackElement *stackElement = stack->top;
	stack->top = stack->top->next;
	delete stackElement;
	return result;
}

bool isEmpty(Stack *stack)
{
	return stack->top == nullptr;
}

StackElement* firstElement(Stack* stack)
{
	return stack->top;
}

StackElement* secondElement(Stack* stack)
{
	return stack->top->next;
}

void deleteStack(Stack *stack)
{
	while (!isEmpty(stack))
	{
		pop(stack);
	}

	delete stack;
}

File: 2 semester/3 homework/Task 3.1/Task 3.1/Stack.cs
using System;

namespace Task_3._1
{
    public class Stack
    {
        class StackElement
        {
            public StackElement pNext { get; private set; }
            public object value { get; private set; }

            public StackElement(object value, StackElement pNext)
            {
                this.pNext = pNext;
                this.value = value;
            }
        }

        private StackElement top;
        private int size;

        public Stack()
        {
            this.top = null;
            this.size = 0;
        }

        /// <summary>
        ///  Добавление элемента
        /// </summary>
        public void Push(object value)
        {
            StackElement newElement = new StackElement(value, this.top);
            this.top = newElement;
            this.size++;
        }

        /// <summary>
        ///  Чтение верхнего элемента
        /// </summary>
        public object Top()
        {
            return this.top.value;
        }

        /// <summary>
        ///  Проверка на пустоту
        /// </summary>
        public bool IsEmpty()
        {
            return this.top == null;
        }

        /// <summary>
        ///  Удаление последнего элемента
        /// </summary>
        public object Pop()
        {
            if (IsEmpty())
            {
                return 1;
            }

            object result = this.top.value;
            this.top = this.top.pNext;
            this.size--;
            return result;
        }

        /// <summary>
        ///  Возвращает размер стека
        /// </summary>
        public int Size()
        {
            return this.size;
        }
    }
}
